### ISSUE TITLE
SpinLocks suffer from false sharing

### ISSUE THREAD

**Issue by kuszmaul** [JuliaLang/julia#55942]
SpinLocks can suffer from false sharing.

**Comment by vtjnash** [comment]
Yes, and many other issues. Don't use them

**Comment by kuszmaul** [comment]
I disagree about `kind:invalid`.   SpinLocks are there. There are cases where they solve a problem that the other locks cannot solve.  It would be easy to provide an option that doesn't suffer from false sharing.  (I have a proposed patch that I'll share).

**Comment by gbaraldi** [comment]
We have a plan to remove them from the language. Their only use should be in places where we can't reschedule stuff i.e the compiler or the scheduler itself.

**Comment by oscardssmith** [comment]
What's wrong with a SpinLock? For sufficiently narrow locks, (e.g. to implement atomics etc), I thought they were relatively close to optimal.

**Comment by kuszmaul** [comment]
Since `SpinLock` is just 8 bytes long, you can end up with two or more of them on the same cache line.

Suppose `lk1` and `lk2` are both on the same cache line and you have two threads waiting on the two locks respectively.  Then the cache line will bounce back and forth between the caches of the two threads. That cache traffic can use up memory bandwidth.

Suppose that the two threads aren't even waiting on the two locks, but just acquire them:
Thread A:
```
@lock lk1 foo1()
bar1()
@lock lk1 zot1()
```
and Thread B:
```
@lock lk2 foo2()
bar2()
@lock lk2 zot2()
```

Here the two threads don't contend at all, logically, but the lock instructions will run slowly because the cache line must bounce back and forth.

This kind of contention is an example of [false sharing](https://en.wikipedia.org/wiki/False_sharing).

**Comment by kuszmaul** [comment]
> We have a plan to remove them from the language. Their only use should be in places where we can't reschedule stuff i.e the compiler or the scheduler itself.

In my experience, sometimes you need a spin lock.  Is there some where I can read about the plan to remove SpinLock?

**Comment by NHDaly** [comment]
@gbaraldi after our discussion on friday I'm still not clear on this. Can you reply to Oscar's and Bradley's questions here?

**Comment by gbaraldi** [comment]
There isn't a plan to remove it. It's just that spin locks in general are quite bad as locks. They are only good when uncontended and held for a very small amount of time

**Comment by alexandergunnarson** [comment]
For the benefit of future readers like myself, posting some latency benchmarks I took. (Warmups prior, core pinning, no GC.) I didn't use the `SpinLock` from this PR, but I did implement several padded spin lock types. The heat maps clearly show the perf comparison between the them and `ReentrantLock`. ([nsync](https://github.com/google/nsync) lock is included for comparison.) @vtjnash I would "contend" (😉) that spin locks certainly have their place in certain scenarios.

Also note that "operation" here is just a simple incrementation of an `Int64`. The same kind of work and amount of work is done by writers and readers.

<img width="6000" height="2000" alt="Image" src="https://github.com/user-attachments/assets/45f40ba5-7767-448e-874c-83f06405df58" />

<img width="6000" height="2000" alt="Image" src="https://github.com/user-attachments/assets/b31a40fd-1440-4e7b-bc3d-519004c90d3b" />

<img width="6000" height="2000" alt="Image" src="https://github.com/user-attachments/assets/0ddcaae0-4441-4ea7-99f8-d69b26d5b163" />

<img width="6000" height="2000" alt="Image" src="https://github.com/user-attachments/assets/01f62757-2807-4118-af35-7ae9a0654437" />

<img width="6000" height="2000" alt="Image" src="https://github.com/user-attachments/assets/4940349a-2d57-49b2-96aa-ec0540ea38f8" />
